### PR TITLE
Add default setting for `Merge matching rests` in Staff properties.

### PR DESCRIPTION
**Pull request by rpatters1**
Resolves: #24983 

Implements a new style setting for the default value of staves' `Merge matching rests` setting.

Screen Shot 2024-10-13 at 6 33 09 AM

The setting in Staff properties is now an `AutoOnOff`:

Screen Shot 2024-10-13 at 6 53 55 AM

- `Auto` takes the setting from the Style setting
- `On` force it on (same as checking the previous setting)
- `Off` forces it off

Since the default value of the new Style setting is unchecked, imports of files created before this Style setting was available remain visually identical. If the staves were unchecked, they are converted to `Auto`. If the staves were checked, they are converted to `On`.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by rpatters1** [comment]
I am changing this to Draft status, because I realized that file imports do not pick up the default. At a minimum, file imports from Midi files and MusicXML files should pick up the default. I can't seem to find a central add staff place that seems safe to add this where all paths would pick it up.

**Comment by rpatters1** [comment]
I added initialization based on the setting to `Factory::createStaff`. This picks up all the imports that I tried without any need to modify them. The only situation where there was a slight issue is calling `staff->init(sourceStaff)`. When the source staff is from the master score and the new staff is in an excerpt, the value put in by `createStaff` could be overwritten with the wrong value. Rather than risk a change to `init`, I simply overrode it with the correct value again in the 2 spots I indentified with this problem.

**Review by cbjeukendrup — COMMENTED** [review]
In this PR, the new style setting is only used to determine the value of "merge matching rests" for newly added staves, and not existing staves. I think that limits the usefulness of this option.

Personally I think it would be preferable to turn the existing staff property into an Auto / On / Off property, where Auto means "follow the score-wide setting". @oktophonie do you think this too?

This could be implemented in two ways:
- Change the type of `Staff::mergeMatchingRests` from `bool` to `AutoOnOff`
- Turn it into a property that is linked to the score-wide style setting; that means that the type stays `bool`, but modifying the value should be done via `getProperty`/`setProperty`, or rather `undoChangeProperty` to make it undoable; `PropertyFlags` are used to determine whether this staff should follow the global style setting (`PropertyFlag::STYLED` means follow the style; `PropertyFlag::UNSTYLED` means the property is overridden for this staff); use `EngravingObject::initElementStyle` in the constructor of `Staff` to set up the "connection".

File: src/engraving/dom/factory.cpp
Comment (rpatters1) [reply src/engraving/dom/factory.cpp:510]: If we change it to AutoOnOff, perhaps we won't a line of code like this anywhere.

**Comment by rpatters1** [comment]
`AutoOnOff` is a much better solution. Thank you for the suggestion. I think it's ready to be reviewed in this format, barring one of the checks failing.

File: src/engraving/rw/write/twrite.cpp
Comment (cbjeukendrup) [inline src/engraving/rw/write/twrite.cpp:2581]: Probably better to use `TConv::toXml(item->mergeMatchingRests())`. Writing enum values by casting to int and then writing that int is risky. Of course, the values of a simple generic enum like AutoOnOff are unlikely to change in practice, but still, better safe than sorry... 
(When writing other AutoOnOff properties, we also use `TConv::toXml`)

File: src/engraving/rw/read400/tread.cpp
Comment (cbjeukendrup) [inline src/engraving/rw/read400/tread.cpp:3775]: In the same spirit as the comment in twrite:
```suggestion
        s->setMergeMatchingRests(e.readInt() ? AutoOnOff::On : AutoOnOff::Off);
```

File: src/engraving/rw/read410/tread.cpp
Comment (cbjeukendrup) [inline src/engraving/rw/read410/tread.cpp:4158]:
```suggestion
        if (ctx.score()->mscVersion() < 450) {
            s->setMergeMatchingRests(e.readInt() ? AutoOnOff::On : AutoOnOff::Off);
        } else {
            s->setMergeMatchingRests(TConv::fromXml(e.readAsciiText(), AutoOnOff::AUTO));
        }
```
You'll also need to update `MSC_VERSION` in `constants.h`; that will need to be done for 4.5 anyway.

File: src/notation/view/widgets/editstaff.ui
Comment (cbjeukendrup) [inline src/notation/view/widgets/editstaff.ui:1079]: This particular deletion should be undone to preserve keyboard accessibility

File: src/notation/view/widgets/editstaff.ui
Comment (cbjeukendrup) [inline src/notation/view/widgets/editstaff.ui:423]: Changes like these are automatically made by recent versions of Qt Creator. They seemed to cause layout issues in other dialogs, so there I had to undo them manually. Perhaps that's not the case here, but it would be good to double-check that there is no weird sizing behaviour when resizing the dialog, also in comparison with the current master branch.

File: src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml
Comment (rpatters1) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:70]: Okay. I fought this hard, being a qml newbie. It resisted many of what seemed the most logical approaches. I will try again.

File: src/notation/internal/masternotation.cpp
Comment (rpatters1) [reply src/notation/internal/masternotation.cpp:361]: If the templates are ever saved in MuseScore 4.5 (with this PR included) then they will have the style setting and override the default. (I tested this, which is why I added this code.) To me, this style setting should override the templates.

Comment (rpatters1) [reply src/notation/internal/masternotation.cpp:361]: It will also override the setting in a user-created template. I guess it's a philosophy of what elements of the user's default style sheet should override what's in the template. I don't have a clear answer in my head, and if you say, "None", then I'll take it out. It isn't as big of a deal now that we are using `AutoOnOff`.

File: src/engraving/rw/read410/tread.cpp
Comment (rpatters1) [reply src/engraving/rw/read410/tread.cpp:4158]: I would like to push back against this. (Because I thought about it carefully.) If we are importing a pre-4.5 file, then `Sid::mergeMatchingRests` (to use your requested name) will be `false`. I think it is a much better import to import staves with the `false` setting to `AUTO`. The net effect will be that matching rests are not merged, but then if the user would like to change them, they don't have to do it one-by-one.

Comment (rpatters1) [reply src/engraving/rw/read410/tread.cpp:4158]: Actually, I am wrong. `Sid::mergeMatchingRests` will be whatever the user has set up in their default style sheet. But I still think setting all the `false` settings to `OFF` is abusive behavior.

Comment (rpatters1) [reply src/engraving/rw/read410/tread.cpp:4158]: I say this because the `false` setting in 4.4 is just there. I'd be willing to bet most users have only given it thought if they need to turn it on. If they've never needed to turn it on, it's just there. Therefore, those staves should be controllable by the Style setting.

Comment (cbjeukendrup) [reply src/engraving/rw/read410/tread.cpp:4158]: Haha, there we go already: I don't know the enum values by heart so had assumed that `Off` would be `0`, but apparently `Off` is `2`. I guess that immediately shows that it's beneficial to make the behaviour more explicit and not rely on casting between enum values and integers.

I have nothing against a variation like
```c++
        if (ctx.score()->mscVersion() < 450) {
            s->setMergeMatchingRests(e.readInt() ? AutoOnOff::ON : AutoOnOff::AUTO);
        } else {
            s->setMergeMatchingRests(TConv::fromXml(e.readAsciiText(), AutoOnOff::AUTO));
        }
```
as long as the behaviour is explicit and clear, and there is no reliance the integer values of the enum.

Comment (rpatters1) [reply src/engraving/rw/read410/tread.cpp:4158]: Should there not be a list of current and past version number constants? I am reluctant to hardcode "450", but if I use Constant::MSC_VERSION it will break when there is a 4.6 (or anything higher than 4.5).

Comment (rpatters1) [reply src/engraving/rw/read410/tread.cpp:4158]: It looks like there is a lot of other numerical hard-coding. Oh well. I've always tried to follow the practice that the only value that can be hard-coded is "0" (and perhaps numbers in well-known math formulae.) But I appreciate that this is an inherited code base.

File: src/notation/view/widgets/editstaff.ui
Comment (rpatters1) [reply src/notation/view/widgets/editstaff.ui:1079]: On macOS I can't get it to tab to any of the comboboxes but perhaps that is a platform issue. (There is a macOS keyboard option to get it to tab to  all controls, but I don't know if these settings control it.) Anyway, I'll change it.

File: src/notation/view/widgets/editstaff.ui
Comment (rpatters1) [reply src/notation/view/widgets/editstaff.ui:423]: I can't see any difference in the resizing behavior between the released 4.4 and the dev branch with these changes. I'll back them out if you wish, but there are several of them.

Comment (rpatters1) [reply src/notation/view/widgets/editstaff.ui:423]: This also could be a platform dependency. I'll try it on Win.

Comment (cbjeukendrup) [reply src/notation/view/widgets/editstaff.ui:423]: Nah, I think it's fine. I also tried it quickly myself, and can confirm that there is no strange behaviour.

File: src/engraving/rw/read410/tread.cpp
Comment (cbjeukendrup) [reply src/engraving/rw/read410/tread.cpp:4158]: I'm planning to introduce a more sophisticated MSC_VERSION system in the near future. For now, hardcoding 450 is probably the best; I'll look for and replace those hardcoded numbers later.

File: src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml
Comment (rpatters1) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:70]: I made the suggested changes, changing `ColumnLayout` to `Column`. If I remove the `anchors` in the `CheckBox`, it looks like this:

Screen Shot 2024-10-16 at 7 34 13 AM

Comment (rpatters1) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:70]: I fought this behavior a long time before I gave up and added the `anchors` originally. 😅

Comment (cbjeukendrup) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:70]: Ah. I think the problem is inside RadioButtonGroup.qml, namely that `implicitHeight` evaluates to `0`. I seem to recall that for some weird reason QML doesn't calculate the `contentHeight` for horizontal list views. 

You can try if removing `implicitHeight: contentHeight` solves anything (perhaps the default value of `implicitHeight` turns out to be the correct one) and otherwise it would be `implicitHeight: Math.max(1, contentItem.childrenRect.height)`.

Comment (cbjeukendrup) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:70]: See https://github.com/musescore/MuseScore/pull/24326 why it is (or rather, will be) necessary.

File: src/notation/internal/masternotation.cpp
Comment (cbjeukendrup) [reply src/notation/internal/masternotation.cpp]: I'm still not very confident about this, but I think we should take it out, because:
- if the user explicitly sets this to true in a template, and then reuses that template, it's reasonable to expect that the value from the template will be taken
- currently, we're not applying this treatment to any other style setting, except to page settings which are of course a bit special. To keep behaviour consistent, I wouldn't be in favour of adding dedicated logic for this specific setting.

**Review by cbjeukendrup — COMMENTED** [review]
Final comments, I think!

**Comment by rpatters1** [comment]
Apparently, changing `constants.h` to `450` will entail a lot of additional changes in test files. That seems like it is significant scope creep for this PR. Can I leave it be for now?

File: src/notation/internal/masternotation.cpp
Comment (rpatters1) [reply src/notation/internal/masternotation.cpp]: If this becomes an issue, I could see a "use template styles" vs "use default styles" switch somewhere as a better, more complete solution.

**Comment by rpatters1** [comment]
I backed out the non-comment change to `contants.h`. That should allow the `run_tests` GitHub action step to complete without throwing an error. It will mean that the dev branch will not correctly read the settings for `AutoOnOff::OFF` in files saved with the dev branch app until `constants.h` is updated to `450`.

If you think the update should be part of the PR, I will need some guidance on what needs to change in the test files so that the `run_tests` don't fail.

Thank you for your due diligence. I am really impressed and grateful at the care you take.

File: src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml
Comment (rpatters1) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml:64]: Whoops, missed this one. Another commit incoming.

**Comment by Jojo-Schmitz** [comment]
Changing MSCVERSION etc. indeed requites basically all test scores to het changed, I'd rather leave that to the core team ;-)
Or to a script, like `for f in $(find . -name '*.mscx' | xargs grep -l 'museScore version="4.40"'); do sed -i 's/museScore version="4.40">/museScore version="4.50">/g' $f; done` (attantion: eintirely untested)

File: src/notation/internal/masternotation.cpp
Comment (rpatters1) [reply src/notation/internal/masternotation.cpp]: Removed it.

File: src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml
Comment (rpatters1) [reply src/notation/qml/MuseScore/NotationScene/internal/EditStyle/RestOffsetSelector.qml]: The few examples I looked at indeed hardcoded their height. I moved the change to `RadioButtonGroup.qml` and it worked (without specifying `height` or `implicitHeight` in `RestOffsetSelector.qml`.

File: src/engraving/tests/scoreutils_tests.cpp
Comment (rpatters1) [reply src/engraving/tests/scoreutils_tests.cpp]: fixed

**Review by cbjeukendrup — APPROVED** [review]
Thanks, looks good to me from a technical point of view!

**Review by zacjansheski — APPROVED** [review]

**Comment by zacjansheski** [comment]
Tested on MacOS 14, Windows 11, Ubuntu 22.04.3. Approved
https://github.com/musescore/MuseScore/issues/24983 FIXED

**Pull request merged**
